### ISSUE TITLE
Add a way to pin variables or steps in streaming mode to allow access to variables that are not sent at every step

### ISSUE THREAD

**Issue by fbudin69500** [ornladios/ADIOS2#1074]
It happens frequently that some data is saved or sent only at certain steps. For example, the geometry of the mesh is often only written on the first step. 
When using a streaming engine like SST, one reader that connects later than when the data was written will miss this important piece of data that maybe is sent only once.
It would be convenient to have a way to pin some steps that are complete (contain all the variables), or individual variables that need to be read on the first step of the processing when a new reader connects.

**Comment by eisenhauer** [comment]
Historically we have thought about such things in the context of ADIOS1, flexpath and streaming data.  Generally the implementation is not as hard as sorting out a reasonable specification and semantics, understanding how those semantics might interact with things like "LatestTimestep" requests from the reader, queue limits on the writer, etc.  At present, properly implementing reasonable attribute semantics will require attributes to be marshaled separately and those bits of data to be treated as "precious", not subject to loss because of queue limits or other operations, required to be provided to late joining readers, etc.  We could consider a mechanism through which an entire timestep might be tagged for similar treatment.

**Comment by JasonRuonanWang** [comment]
I have been thinking about exactly this issue when I write the new staging engine. It caused problem to me when I was doing loose coupling with attributes. 

It looks to me that currently there isn't a public API allowing us to do this. Once the public API supports this, it's not difficult to set it to always hold certain data in writers, at least in the new staging engine that I am currently writing. Of course you need enough memory to buffer this data.

**Comment by williamfgc** [comment]
Why not putting the information at every step if that's the intended IO behavior pattern anyways. Ultimately, it's either implemented by your workflow (which is supported today) or by the library (not supported today, through an operator in the public API). The end result is the same, duplicate the variable in buffer at every step. We are trading calls to Put by calls to Operator. My 2 cents is that we have a bigger fish to fry and @fbudin69500 has a solution he can already use today.

**Comment by JasonRuonanWang** [comment]
The end result is the same if you don't care performance. Otherwise it's slightly different. If you do it at the application layer then you need to call put and it does serialization underneath for every step. If you provide an operator to hold the variable buffer then you only do it once.

**Comment by williamfgc** [comment]
@JasonRuonanWang you still have to stitch the variable index and data buffer with the current step buffer. Not only that, but test it under several scenarios which don't work today. By bigger fish I mean the release, and that's not up to me.

**Comment by eisenhauer** [comment]
A new feature has been added to SST.  There is an engine parameter, FirstTimestepPrecious that causes the first submitted typestep to be essentially saved forever and provided as the first timestep to readers that might join later in the workflow...  This only works for the SST engine (at the moment).  Documentation is here: 
https://adios2.readthedocs.io/en/latest/engines/engines.html#sst-sustainable-staging-transport
@fbudin69500, please let us know if this functionality addresses your needs...

**Comment by eisenhauer** [comment]
I'm going to go ahead and close this.  If there are needs that aren't covered by FirstTimestepPrecious (and there may well be), please open another ticket to describe them.
